packages/viewer/src/scenes/operations.ts

```diff
@@ -0,0 +1,241 @@
+    return new AndSelectorBuilder(this.parent);
+  }
+
+  public build(): OrSelector {
+    return { type: 'or', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * A selector builder to perform boolean `and` operations.
+ */
+export class AndSelectorBuilder implements SelectorBuilder<AndSelector> {
+  private builders?: SelectorBuilder<ItemSelector>[] = [];
+  private parent: SelectorBuilder<ItemSelector>;
+  public constructor(parent: SelectorBuilder<ItemSelector>) {
+    this.builders.push(parent);
+    this.parent = parent;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return new OrSelectorBuilder(this.parent);
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return this;
+  }
+
+  public build(): AndSelector {
+    return { type: 'and', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * An `ItemSelector` describes a type for performing a selection against some
+ * attribute of a scene item.
+ */
+export type ItemSelector = SuppliedIdSelector | ItemIdSelector;
+
+interface OrSelector {
+  type: 'or';
+  selectors: ItemSelector[];
+}
+
+interface AndSelector {
+  type: 'and';
+  selectors: ItemSelector[];
+}
+
+export interface OperationDefinition {
+  operation: ItemOperation | SceneOperation;
+}
+
+export interface SelectorBuilder<T> {
+  /**
+   * Returns the built selector.
+   */
+  build(): T;
+
+  /**
+   * Returns a conditional builder to perform an `or` operation.
+   */
+  or(): SelectorBuilder<OrSelector>;
+
+  /**
+   * Returns a conditional builder to perform an `or` operation.
+   */
+  and(): SelectorBuilder<AndSelector>;
+
+  /**
+   * Selects a item that has the given Customer facing ID
+   *
+   * @param suppliedId The customer ID of an item to match
+   */
+  withSuppliedId(suppliedId: string): this;
+
+  /**
+   * Selects a item that has the given Vertex ID.
+   *
+   * @param itemId The ID of the item to match.
+   */
+  withItemId(itemId: string): this;
+}
+
+interface ChangeMaterialOperation {
+  type: 'change-material';
+  color: string;
+}
+
+export type SceneOperation = ChangeMaterialOperation;
+
+export type ItemOperation = ShowItemOperation | HideItemOperation;
+
+export interface SceneItemQuery {
+  where(
+    query: SelectorBuilder<ItemSelectorBuilder>
+  ): SceneItemOperationsExecutor;
+}
+
+export interface SceneItemOperations<T> {
+  material(color: string): T;
```

Comment:
Let's rename this to `materialOverride`. Also, I think we should introduce a type that represents the material. This way we can extend this to support different types of materials in the future. There's other fields on our ColorMaterial too, so having a separate type would allow others to be specific in what gets updated. See https://github.com/Vertexvis/vertex-api/blob/master/src/main/resources/schemas/ColorMaterial.yml for how this is defined in vertex-api.

We could also have a helper `ColorMaterial.fromHex()` which just sets the diffuse field.